### ISSUE TITLE
shader float bit widths: restrict to 32 bits

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#230]
In vanilla Vulkan, floating point scalars must be 32-bit.
shaderFloat64 is an optional feature in Vulkan 1.0
shaderFloat16 is an optional feature introduced by extension VK_KHR_float16_int8

For widest availability, the WebGPU shader programming model should be restricted to 32-bit floats.

**Comment by dneto0** [comment]
Cc: @RobinMorisset

**Comment by dneto0** [comment]
See also #229 for the integer case.

The argument from https://github.com/gpuweb/WHLSL/issues/165 can be adapted here:
You might expose 16bit and 64bit floats in the programming model and require the implementation to emulate it if it's not natively supported.

The costs might be quite high, enough that we might not want to expose it for that reason.

**Comment by Kangz** [comment]
+1 for not exposing it if it is expensive to emulate. I think FP16 support will be a very good extension once WebGPU is out.

**Comment by RobinMorisset** [comment]
I agree that we should not mandate things that are very expensive to emulate. This rules out 64-bits floats (doubles).

I don't have a strong opinion on the cost/usefulness of supporting 16-bit floats. Not supporting them in the MVP and making them an early extension sounds reasonable.

**Comment by dneto0** [comment]
Some more background on this issue:

FP16 support can be split into two parts:

- memory access of FP16 scalars and vectors
- arithmetic done in FP16

Some implementations have native support for one and not the other.

Example: 
- memory access without arithmetic:
  - Vulkan and SPIR-V introduced an extension for "16 bit storage" ( e.g. http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_16bit_storage.html ) which enables memory access for aligned 16bit values.  But as soon as you load such a value you immediately convert to 32-bit quantities for arithmetic. And the output of a narrowing conversion may only be used as the value for a subsequent store.

- FP16 arithmetic without 16bit memory access
  - Two (non-atomic) memory accesses to the two halves of a 32-bit word will race.  So raw "half" memory accesses are not immediately supportable.
  - I believe such an implementation can support memory accesses of half2, half3, half4.
  - To emulate scalar "half" accesses, you could use loop-around-atomic-cmpexch, but I don't know that progress guarantees exist to ensure that works.  There is also a performance penalty which can count into the tradeoff.

A point about emulation of FP16 arithmetic by FP32 operations.  Yes, it's trivially doable, but *edge case behaviour is different*, especially overflow of intermediates.    Example is the `normalize(half4 v)` builtin.  It is essentially   `  v / length(v)` or ` v / sqrt(dot(v,v))`.  If the intermediates are implemented in FP32, then you can get a large dot-product (bigger than FP16 max value), then the sqrt reins it back into FP16 normal range.    So if emulated with FP32 intermediates you an get a finite answer, but if implemented truly in FP16 then you get an infinity and quite bad behaviour.
This is a real and subtle portability problem.  It's not clear what the right tradeoff is here at the spec level, given that only the programmer knows the domain they're working in and knows how much they want performance vs. the portability.

**Comment by grorg** [comment]
Discussed at the [8 April 2019 meeting](https://docs.google.com/document/d/1UiIZ7G7iruGYntiLSmhc-wRPj1sN1CqNcPdgeG51r1U/edit)

**Comment by grorg** [comment]
Resolution: 32 bit floats for MVP.

Extensions for maybe 16 and 64 can come later.
